Pioneer: Fix for #511

mnaamani opened the pull request:
Attempt to Fix #511

mnaamani committed:
pioneer: some import fixes in joy-utils

mnaamani commented:
I'm convinced now there is something uniquely wrong with my local dev environment. Clearly I have some cached typescript definitions files interfering with the build

mnaamani commented on pioneer/packages/joy-utils/src/react/context/transport.tsx:
```diff
@@ -2,7 +2,7 @@ import React, { createContext, useContext } from "react";
 import { ApiProps } from "@polkadot/react-api/types";
 
-import Transport from "../../transport";
+import { Transport } from "../../transport";
```
I can clearly see by examining `../../transport/index.ts` that there is a default export class statement which means this fix is incorrect, but I can't figure out how my IDE is not seeing the same thing :)

<img width="1046" alt="Screen Shot 2020-05-29 at 9 19 48 PM" src="https://user-images.githubusercontent.com/1621012/83292055-23066280-a1f2-11ea-92e8-ca1176321174.png">

Lezek123 replied:
I think I know what's the issue here. There is `joy-utils/src/Transport` (file) and `joy-utils/src/transport` (directory). Perhaps in some environments where paths are considered case insensitive import from file takes priority over the import from directory and hence the error.

mnaamani commented:
I think you are right on the money! OSX has strange hybrid case sensitivity
and this is failing on my Mac.

Not the first time we’ve had issues with case sensitive issues with file
names :(

On Fri, 29 May 2020 at 10:23 pm, Lezek123 <notifications@github.com> wrote:

> *@Lezek123* commented on this pull request.
> ------------------------------
>
> In pioneer/packages/joy-utils/src/react/context/transport.tsx
> <https://github.com/Joystream/joystream/pull/514#discussion_r432688301>:
>
> > @@ -2,7 +2,7 @@ import React, { createContext, useContext } from "react";
>  import { ApiContext } from "@polkadot/react-api";
>  import { ApiProps } from "@polkadot/react-api/types";
>
> -import Transport from "../../transport";
> +import { Transport } from "../../transport";
>
> I think I know what's the issue here. There is joy-utils/src/Transport
> (file) and joy-utils/src/transport (directory). Perhaps in some
> environments where paths are considered case insensitive import from file
> takes priority over the import from directory and hence the error.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/Joystream/joystream/pull/514#discussion_r432688301>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAMLYFBFZOCH5RNDIBHFLMTRUADTPANCNFSM4NOHN4IA>
> .
>

mnaamani committed:
pioneer: joy-utils fix case-(in)-sensitivity issue with import path

On Mac is seems node import paths files take precedence over folder because of lack of
real case sensitivity in OSX file system. eg in this case: Transport.ts and ./transport (folder)

Added macos to github actions matrix to test and confirm

mnaamani committed:
github workflows: separate jobs per runner

mnaamani committed:
github workflows: fix bad syntax in pioneer-pr.yml

mnaamani committed:
github workflows: last syntax fix, reduce jobs

mnaamani commented:
So the quick and dirty fix 2f2df78 confirms it is the issue you pointed out. I'm not sure there is an environment var to control the precedence of how node handles import paths when a file or folder match, but in this particular case its not really the issue, its actually the fact that it can't tell the different between `transport.ts` and `Transport.ts` and `./transport/` due to OSX filesystem case sensitivity.

I added a job to github workflows to build for OSX so we can spot such intricate "bugs" in future, but it would be better I think to follow a convention where we try to avoid having a file and folder with similar names that would cause this case to be problematic.

Lezek123 reviewed (approved):
The CI works well, I was able to reproduce the error in my fork.
This simple fix seems to be enough for now, I will rename the file or change the structure in this directory in some PR later (`joy-utils` needs some cleanup anyway, as it is currently partially the old `joy-utils` and patrially a work-in-progress `joystream-js`)
